integration_test/support/mix_task_runner.exs

```diff
@@ -0,0 +1,170 @@
+defmodule Phoenix.Integration.MixTaskRunner do
+  @moduledoc """
+  This runs mix tasks via the command-line.
+  This is useful for running tasks in projects outside of the
+  current BEAM instance.
+  """
+
+  use GenServer
+
+  @type prompt_response :: :no_to_all | :yes_to_all | String.t() | [String.t()]
+  @type opt ::
+          {:prompt_responses, [prompt_response]}
+          | {:cd, String.t()}
+          | {:env, [{String.t(), String.t()}]}
+
+  @type args :: [String.t()]
+
+  @spec start_link(args, [opt]) :: GenServer.on_start()
+  def start_link(args, opts) when is_list(args) and is_list(opts) do
+    parent_pid = self()
+    GenServer.start_link(__MODULE__, {parent_pid, args, opts})
+  end
+
+  @spec run(args, [opt]) :: {String.t(), exit_status :: non_neg_integer()}
+  def run(args, opts \\ []) when is_list(args) and is_list(opts) do
+    {:ok, pid} = start_link(args, opts)
+
+    receive do
+      {:unexpected_prompt, ^pid, {prompt, all_output}} ->
+        raise """
+        unexpected prompt
+
+          #{prompt}
+
+        All output:
+
+        #{all_output}
+        """
+
+      {:command_exited, ^pid, {output, exit_code}} ->
+        {output, exit_code}
+    end
+  end
+
+  def run!(args, opts) when is_list(args) and is_list(opts) do
+    case run(args, opts) do
+      {output, 0} ->
+        output
+
+      {output, exit_code} ->
+        raise """
+        mix command failed with exit code: #{inspect(exit_code)}
+
+        mix #{Enum.join(args, " ")}
+
+        #{output}
+
+        Options
+          cd: #{get_cd(opts)}
+          env: #{inspect(get_env(opts))}
+        """
+    end
+  end
+
+  @impl true
+  def init({parent_pid, args, opts}) do
+    mix_path = System.find_executable("mix")
+    port_options = extract_port_options(opts)
+
+    prompt_responses =
+      opts
+      |> Keyword.get(:prompt_responses, [])
+      |> List.wrap()
+
+    port =
+      Port.open(
+        {:spawn_executable, mix_path},
+        [:exit_status, :stderr_to_stdout, args: args, env: get_env(opts)] ++ port_options
+      )
```

Comment:
Can we use `System.cmd` instead of Port.open? 🤔

Reply:
Actually, I see you want to have control over the response... I am honestly a bit torn. Shelling out to start multiple commands are going to make the integration tests very expensive. I would prefer if we could invoke Mix directly from the current instance, without starting a new app, and use Mix.Shell.Process. It makes the tests less of integration tests but I think the benefits will be visible in the long term. It is a difference between commands taking 2ms vs 200ms. We do already have a structure similar to what I described in `phx.new_test` which maybe we can extend to support these use cases?

Thoughts?

Reply:
I started with the approach that phoenix currently uses of running the new app inside of the current beam instance, but ran into a few problems.

1. Mix is designed to be a singleton, so running tests with `async: true` will cause unexpected errors. The tests in this suite are about 30 seconds faster on my machine when running `mix test.integration` (async) instead of `mix test.integration --trace`
1. I had difficulty with checks like [this one](https://github.com/aaronrenner/phx_gen_auth/blob/1a8e74deb4eff273dc60b16b64dbee1ad701d91e/lib/mix/tasks/phx.gen.auth.ex#L155-L157) when running in the same beam instance as other tests that do load these libraries.
    ```elixir
    unless Code.ensure_loaded?(Ecto.Adapters.SQL) do
      raise_with_help("mix phx.gen.auth requires ecto_sql", :phx_generator_args)
    end
    ```
1. I believe there's several places where I've seen settings from phoenix itself affect how the generated app runs, and had to keep that in mind or work around it when going through the tests.

Overall, I think we'll run into less issues in the long run if we run these tasks as a separate beam instance. I think we'd get the most performance improvements is to come up with a caching strategy for deps and the compiled deps, such that each dependency only has to be compiled once for the entire test run (instead of once per app). I've made some attempts at this with phx_gen_auth but still haven't come up with an ideal solution.

Reply:
Thank you for the reply. I agree with your points then! One last question then: is it possible for us to use `System.cmd` and abandon the prompts? Instead we can use flags such as `--force` to always run particular commands in certain ways. WDYT?

Also note that, once this is in, we should consider removing the `phx.new_test` integration tests that we have today, once we have fully ported them.

Reply:
At the moment, the need to respond "No" when generating a new project and asked if we want to fetch and install dependencies. Are you thinking we'd add a `mix phx.new --force` option to not install dependencies?

https://github.com/phoenixframework/phoenix/blob/0c045af1a86b24cfa02187abdeeb58da6cd6dcaf/installer/lib/mix/tasks/phx.new.ex#L150-L154

Here are the other places where the generators ask for responses:


_Prompts if the user wants to overwrite the directory_
https://github.com/phoenixframework/phoenix/blob/0c045af1a86b24cfa02187abdeeb58da6cd6dcaf/installer/lib/mix/tasks/phx.new.ex#L346-L350

_Asks the user if they want to proceed when generating into an existing context_
https://github.com/phoenixframework/phoenix/blob/0c045af1a86b24cfa02187abdeeb58da6cd6dcaf/lib/mix/tasks/phx.gen.context.ex#L257-L283

_Prompts the user if they want to overwrite existing files_
https://github.com/phoenixframework/phoenix/blob/0c045af1a86b24cfa02187abdeeb58da6cd6dcaf/lib/mix/phoenix.ex#L318-L342

I do have a test in phx_gen_auth that generates into an existing context and has to respond yes to a prompt: https://github.com/aaronrenner/phx_gen_auth/blob/1a8e74deb4eff273dc60b16b64dbee1ad701d91e/test/mix/tasks/phx_gen_auth/integration_tests/default_app_test.exs#L129. This module is also nice because it will raise if it encounters an unexpected prompt instead of the test just hanging.

I'm completely happy to remove this module if you think there's a better way to do this, but we'll have to come up with a way to bypass these prompts.

Reply:
The first one can be solved with `--skip-dependencies` that we will add. I would solve all other prompts with `--force`. :)

Reply:
That simplifies things a lot. :smile:  I'll submit a PR for that.

Reply:
Sounds great! Now that I think about it, I would probably call the flag `--install` and declare it as a boolean, so we get `--no-install` for free. :) Also please directly ping me once you submit the new PR so I take a look at the code, as I don't receive notifications for this repo.